### PR TITLE
(CAT-1301) Add check unsafe interpolations check

### PR DESCRIPTION
**Pull request by GSPatton**
This commit adds the check unsafe interpolations check to core puppet-lint insted of having it as a standalone gem

## Summary
Provide a detailed description of all the changes present in this pull request.

## Additional Context
Add any additional context about the problem here. 
- [ ] Root cause and the steps to reproduce. (If applicable)
- [ ] Thought process behind the implementation.

## Related Issues (if any)
Mention any related issues or pull requests.

## Checklist
- [ ] 🟢 Spec tests.
- [ ] 🟢 Acceptance tests.
- [ ] Manually verified.

### CODE REVIEW

File: lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  # Finds the index offset of the next instance of `value` in `tokens_slice` from the original index
+  def index_offset_for(value, tokens_slice)
+      i = 0
+      while i < tokens_slice.length
+      return i if value.include?(tokens_slice[i].value)
+      i += 1
+      end
+  end
+end
+
```
Comment (ekohl) [inline lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:130]:
```suggestion
```

**Review by ekohl — CHANGES_REQUESTED** [review]
This does depend on PUP-5704, so that raises the minimum compatibility to Puppet 7.10.0 which should be noted in the compatibility section of README.

And RuboCop is not happy with you.

File: lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb (hunk 1 above)
Comment (GSPatton) [reply lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:130]: wow, Rubocop really isn't happy with me... thanks @ekohl. I'll fix this up

Comment (GSPatton) [reply lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:130]: also, what is PUP-5704? Do you mean a specific version of puppet?

File: lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  end
+
+  # Finds the index offset of the next instance of `value` in `tokens_slice` from the original index
+  def index_offset_for(value, tokens_slice)
+    i = 0
+    while i < tokens_slice.length
+      return i if value.include?(tokens_slice[i].value)
+
+      i += 1
+    end
```
Comment (ekohl) [inline lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:130]: Why not:
```suggestion
    tokens_slice.each_with_index do |token, i|
      return i if value.include?(token.value)
    end
```

File: lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  # This function adds a check for titles in double quotes where there could be interpolated variables
+  def get_exec_titles
+    result = []
+    tokens.each_index do |token_idx|
+      next unless ['exec'].include?(tokens[token_idx].value)
+
+      # We have a resource declaration. Now find the title
+      tokens_array = []
+      # Check if title is an array
+      if tokens[token_idx].next_code_token.next_code_token.type == :LBRACK
```
Comment (ekohl) [inline lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:91]: I'd use safe navigators. If there is no next token (which can happen on files with invalid syntax) your program won't crash.
```suggestion
      if tokens[token_idx]&.next_code_token&.next_code_token&.type == :LBRACK
```

File: lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+      current_token = current_token.next_token
+    end
+  end
+
+  # This function is a replacement for puppet_lint's title_tokens function which assumes titles have single quotes
+  # This function adds a check for titles in double quotes where there could be interpolated variables
+  def get_exec_titles
+    result = []
+    tokens.each_index do |token_idx|
+      next unless ['exec'].include?(tokens[token_idx].value)
```
Comment (ekohl) [inline lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:86]: I think this is easier to read
```suggestion
      next if tokens[token_idx].value != 'exec'
```

File: lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+      current_token = current_token.next_token
+    end
+  end
+
+  # This function is a replacement for puppet_lint's title_tokens function which assumes titles have single quotes
+  # This function adds a check for titles in double quotes where there could be interpolated variables
+  def get_exec_titles
+    result = []
+    tokens.each_index do |token_idx|
```
Comment (ekohl) [inline lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:85]: This is a bit inefficient. Why not iterate with an index so you don't need to look up the value every time?
```suggestion
    tokens.each_with_index do |token, token_idx|
```

File: lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+    rule_violations
+  end
+
+  # A command is parameterised if its args are placed in an array
+  # This function checks if the current token is a :FARROW and if so, if it is followed by an LBRACK
+  def parameterised?(token)
+    current_token = token
+    while current_token.type != :NEWLINE
+      return true if current_token.type == :FARROW && current_token.next_token.next_token.type == :LBRACK
```
Comment (ekohl) [inline lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:75]: I don't think this is valid. There may be multiple tokens between, like whitespace.

File: lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    end
+
+    rule_violations
+  end
+
+  # A command is parameterised if its args are placed in an array
+  # This function checks if the current token is a :FARROW and if so, if it is followed by an LBRACK
+  def parameterised?(token)
+    current_token = token
+    while current_token.type != :NEWLINE
```
Comment (ekohl) [inline lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:74]: Puppet doesn't care about whitespace, so why does this matter? Don't you want to look for a comma or `}` instead, which signifies that the next attribute comes or the definition is over.

File: lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb [hunk 8]
```diff
@@ -0,0 +1,132 @@
+PuppetLint.new_check(:check_unsafe_interpolations) do
+  COMMANDS = Array['command', 'onlyif', 'unless']
+  INTERPOLATED_STRINGS = Array[:DQPRE, :DQMID]
+  USELESS_CHARS = Array[:WHITESPACE, :COMMA]
+  def check
+    # Gather any exec commands' resources into an array
+    exec_resources = resource_indexes.map { |resource|
```
Comment (ekohl) [inline lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb:7]: Ruby 2.7 has `filter_map` which is `.map.compact`.

**Review by ekohl — COMMENTED** [review]
I feel like this is one check that would work must better on a parsed tree, rather than on tokens. Today puppet-lint doesn't offer such a thing, but it may be worth considering if more of these checks are added.

File: .rubocop_todo.yml [hunk 9]
```diff
@@ -14,6 +14,41 @@ Gemspec/RequireMFA:
   Exclude:
     - 'puppet-lint.gemspec'
 
+# Offense count: 1
```
Comment (ekohl) [inline .rubocop_todo.yml:17]: Why add these as commented lines?

File: .rubocop_todo.yml [hunk 10]
```diff
@@ -92,13 +132,19 @@ Performance/CollectionLiteralInLoop:
     - 'lib/puppet-lint/plugins/check_resources/ensure_first_param.rb'
     - 'lib/puppet-lint/plugins/check_whitespace/trailing_whitespace.rb'
 
-# Offense count: 406
+# Offense count: 1
+# This cop supports unsafe autocorrection (--autocorrect-all).
+Performance/MapCompact:
+  Exclude:
+    - 'lib/puppet-lint/plugins/check_unsafe_interpolations/check_unsafe_interpolations.rb'
```
Comment (ekohl) [inline .rubocop_todo.yml:139]: Looks like you already fixed this, so perhaps this is gone if you regenerate the TODO file.

**Review by david22swan — APPROVED** [review]

File: .rubocop_todo.yml (hunk 10 above)
Comment (GSPatton) [reply .rubocop_todo.yml]: good spot! Thanks

**Comment by ekohl** [comment]
This does give false positives. If command is present the title doesn't need to be validated. An example:
https://github.com/voxpupuli/puppet-trusted_ca/blob/8f9ccc04a12ed3608131d9cb4d80d9d9ecffb380/manifests/ca.pp#L74-L81

Or simplified:
```puppet
exec { "validate ${filename}":
  command => "openssl x509 -in ${filename} -noout",
}
```

**Pull request merged**